src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,733 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
+    pub fn value(&self) -> &D::Type {
+        self.data.as_value()
+    }
+
+    /// Returns a mutable reference to register data.
+    pub fn value_mut(&mut self) -> &mut D::Type {
+        self.data.as_mut_value()
+    }
 
-        set_registers(&vcpu, &regs).unwrap();
-        let off = offset__of!(user_pt_regs, pstate);
-        let id = arm64_core_reg_id!(KVM_REG_SIZE_U64, off);
-        let pstate = vcpu
-            .get_one_reg(id)
-            .expect("Failed to call kvm get one reg");
-        assert!(regs.contains(&Aarch64Register { id, value: pstate }));
+    /// Returns register data as byte slice
+    pub fn as_slice(&self) -> &[u8] {
+        self.data.as_slice()
     }
 
+    /// Returns register data as mutable byte slice
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
+        self.data.as_mut_slice()
+    }
+}
+
+/// Trait for types that can represent register data
+pub trait RegData {
+    /// Type of value that can be represented
+    type Type;
+    /// Constructs new `RegData`
+    fn new(data: Self::Type) -> Self;
+    /// Return reference to data
+    fn as_value(&self) -> &Self::Type;
+    /// Return mutable reference to data
+    fn as_mut_value(&mut self) -> &mut Self::Type;
+    /// Returns data as a byte slice
+    fn as_slice(&self) -> &[u8];
+    /// Returns data as a mutable byte slice
+    fn as_mut_slice(&mut self) -> &mut [u8];
+}
+
+macro_rules! reg_data {
+    ($name:ident, $t:ty) => {
+        /// Register data type
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        pub struct $name {
+            data: $t,
+        }
+
+        impl Default for $name {
+            fn default() -> Self {
+                Self { data: 0 }
+            }
+        }
+
+        impl RegData for $name {
+            type Type = $t;
+
+            fn new(data: Self::Type) -> Self {
+                Self { data }
+            }
+
+            fn as_value(&self) -> &Self::Type {
+                &self.data
+            }
+
+            fn as_mut_value(&mut self) -> &mut Self::Type {
+                &mut self.data
+            }
+
+            #[allow(clippy::size_of_in_element_count)]
+            fn as_slice(&self) -> &[u8] {
+                // # Safety:
+                // Slice over known type
+                unsafe {
+                    std::slice::from_raw_parts(
+                        (&self.data as *const $t).cast::<u8>(),
+                        std::mem::size_of::<Self::Type>(),
+                    )
+                }
+            }
+
+            #[allow(clippy::size_of_in_element_count)]
+            fn as_mut_slice(&mut self) -> &mut [u8] {
+                // # Safety:
+                // Slice over known type
+                unsafe {
+                    std::slice::from_raw_parts_mut(
+                        (&mut self.data as *mut $t).cast::<u8>(),
+                        std::mem::size_of::<Self::Type>(),
+                    )
+                }
+            }
+        }
+    };
+}
```

Comment:
This looks to me like a reimplementation of `ByteValued` from vm-memory - can we use that trait?

Reply:
Simplified implementation. Now `RegData#` types just have a byte arrays inside.